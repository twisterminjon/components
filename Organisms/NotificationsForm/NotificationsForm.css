.notificationsform {
  display: flex;
  flex-direction: column;

  color: var(--menu__fg);
  background-color: var(--menu__bg);

  width: 100%;
  height: 100%;
}

.notificationsform-title {
  font-size: 22px;
  font-weight: bold;

  margin: 23px 0px 0 0px;
}

.notificationsform-section {
  padding: 26px 16px;
}

.notificationsform-section--last {
  /* Fill the remaining space */
  flex: 1;

  /* push content to bottom */
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
}

.notificationsform hr {
  border-top: 1px solid var(--brandcolor_muted);
  border-bottom: none;
  margin-left: -16px;
  margin-right: -16px;
}
