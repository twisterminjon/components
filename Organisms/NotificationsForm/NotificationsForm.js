import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import dayjs from 'dayjs'

import { DateFormat, ProjectDate } from '@shared/helpers'

import Button from '../../Atoms/Button/Button'
import MenuItem from '../../Molecules/MenuItem/MenuItem'

import './NotificationsForm.css'
import NotificationsPauseToast from '../../Molecules/NotificationsPauseToast/NotificationsPauseToast'

export default class NotificationsForm extends Component {
  static propTypes = {
    /** Called after the close button is clicked */
    onClose: PropTypes.func.isRequired,

    /** Called after notifications are paused  */
    onPause: PropTypes.func.isRequired,

    /** Set to true if notifications are paused */
    paused: PropTypes.bool,

    /** Called after notifications are resumed */
    onResume: PropTypes.func.isRequired,

    /** DateTime when notifications should be paused */
    defaultPauseStartDateTime: PropTypes.string,

    /** DateTime when notifications should be unpaused */
    defaultPauseEndDateTime: PropTypes.string,

    /** Whether notifications are scheduled to be paused */
    scheduled: PropTypes.bool,
  }

  static defaultProps = {
    paused: false,
    defaultPauseStartDateTime: '',
    defaultPauseEndDateTime: '',
    scheduled: false,
  }

  constructor(props) {
    super(props)

    this.state = {
      showingNotificationsPauseToast: false,
    }
  }

  /**
   * Shows or hides the custom time picker.
   *
   * @param {boolean} isShowing
   */
  handleShowNotificationPauseToast(isShowing = true) {
    this.setState({
      showingNotificationsPauseToast: isShowing,
    })
  }

  handlePause = ({ startDateTime, endDateTime }) => {
    const { onPause } = this.props

    if (typeof onPause === 'function') {
      onPause({
        startDateTime,
        endDateTime,
      })
    }
  }

  /**
   * Creates a time range from the current dayjs to the end dayjs
   *
   * @param {dayjs.Moment} endMoment? [optional]
   * @return {Object}
   */
  createPauseTimeRange(endMoment = null) {
    const startDateTime = ProjectDate().now(DateFormat.DateTime3)
    const endDateTime = endMoment ? endMoment.format(DateFormat.DateTime3) : null

    return {
      startDateTime,
      endDateTime,
    }
  }

  render() {
    const {
      onPause,
      paused,
      onResume,
      defaultPauseStartDateTime,
      defaultPauseEndDateTime,
      onClose,
      scheduled,
    } = this.props
    const { showingNotificationsPauseToast } = this.state

    // Menu options to turn on pausing
    const pauseMenuOptions = [
      {
        name: 'until-off',
        label: 'Until I turn off',
        onClick: () => onPause(this.createPauseTimeRange(null)),
      },
      {
        name: '30-minutes',
        label: '30 minutes',
        onClick: () => onPause(this.createPauseTimeRange(dayjs().add(30, 'minutes'))),
      },
      {
        name: '1-hour',
        label: '1 hour',
        onClick: () => onPause(this.createPauseTimeRange(dayjs().add(1, 'hour'))),
      },
      {
        name: '2-hours',
        label: '2 hours',
        onClick: () => onPause(this.createPauseTimeRange(dayjs().add(2, 'hours'))),
      },
      {
        name: '4-hours',
        label: '4 hours',
        onClick: () => onPause(this.createPauseTimeRange(dayjs().add(4, 'hours'))),
      },
      {
        name: 'until-tomorrow',
        label: 'Until tomorrow',
        onClick: () => onPause(this.createPauseTimeRange(dayjs().add(1, 'day').startOf('day'))),
      },
      {
        name: 'custom',
        label: 'Custom',
        onClick: () => this.handleShowNotificationPauseToast(true),
      },
    ]

    // Menu options to turn off, or adjust, pausing
    const unpauseMenuOptions = [
      {
        name: 'turn-on-notifications',
        label: 'Turn notifications back on',
        onClick: onResume,
      },
      {
        name: 'adjust-time',
        label: 'Adjust the time',
        onClick: () => this.handleShowNotificationPauseToast(true),
      },
    ]

    const menuOptions = !paused ? pauseMenuOptions : unpauseMenuOptions

    return (
      <div className="notificationsform" data-testid="notificationsform">
        <div className="notificationsform-title notificationsform-section">Notifications</div>
        <hr />
        <div className="notificationsform-section">
          {paused ? (
            <Fragment>
              <p>Your notifications are paused.</p>
              <p>Would you like to</p>
            </Fragment>
          ) : (
            <Fragment>
              <p>Pause your notifications.</p>
              <p>When your notifications are paused other users will be notified that you are unavailable.</p>
            </Fragment>
          )}
        </div>
        <hr />
        <div>
          {menuOptions.map(({ name, label, onClick }, idx) => (
            <MenuItem
              data-testid={`notificationsform-menu-item-${label}`}
              key={idx}
              name={name}
              onClick={onClick}
              label={label}
            />
          ))}
        </div>
        {scheduled && (
          <Fragment>
            <hr />
            <MenuItem
              data-testid={`notificationsform-menu-item-adjust-scheduled-pause-time`}
              name={'adjust-scheduled-pause-time'}
              onClick={() => this.handleShowNotificationPauseToast(true)}
              label="Adjust scheduled pause time"
            />
            <MenuItem
              data-testid={`notificationsform-menu-item-cancel-schedule`}
              name={'cancel-schedule'}
              onClick={onResume}
              label="Cancel scheduled pause"
            />
          </Fragment>
        )}
        <div className="notificationsform-section notificationsform-section--last">
          <Button data-testid="notificationsform-close-button" size="big" fluid onClick={onClose}>
            Close
          </Button>
        </div>

        {showingNotificationsPauseToast && (
          <NotificationsPauseToast
            show={true}
            defaultStartDateTime={defaultPauseStartDateTime}
            defaultEndDateTime={defaultPauseEndDateTime}
            onPause={pauseData => {
              this.handleShowNotificationPauseToast(false)
              this.handlePause(pauseData)
            }}
            onCancel={() => this.handleShowNotificationPauseToast(false)}
          />
        )}
      </div>
    )
  }
}
