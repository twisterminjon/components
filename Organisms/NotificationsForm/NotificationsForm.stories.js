import React from 'react'

import { storiesOf } from '@storybook/react'
import StoryRouter from 'storybook-react-router'

import { withKnobs, boolean, text } from '@storybook/addon-knobs'

import NotificationsForm from './NotificationsForm'

const stories = storiesOf('Provider/v1/Organisms', module)
stories.addDecorator(withKnobs)

stories.addDecorator(StoryRouter()).add('NotificationsForm', () => {
  const showLabel = 'show'
  const showDefaultValue = true
  const showValue = boolean(showLabel, showDefaultValue)

  const pausedLabel = 'paused'
  const pausedDefaultValue = false
  const pausedValue = boolean(pausedLabel, pausedDefaultValue)

  const scheduledLabel = 'scheduled'
  const scheduledDefaultValue = false
  const scheduledValue = boolean(scheduledLabel, scheduledDefaultValue)

  const defaultPauseStartDateTimeValue = text('defaultPauseStartDateTimeValue', '2020-03-04T07:45')
  const defaultPauseEndDateTimeValue = text('defaultPauseEndDateTimeValue', '2020-03-04T09:45')

  return (
    <div
      style={{
        position: 'absolute',
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
      }}>
      <NotificationsForm
        show={showValue}
        paused={pausedValue}
        onPause={({ startDateTime, endDateTime }) => alert(`Paused from ${startDateTime} to ${endDateTime}`)}
        onResume={() => alert('onResume')}
        onClose={() => alert('onClose')}
        defaultPauseStartDateTime={defaultPauseStartDateTimeValue}
        defaultPauseEndDateTime={defaultPauseEndDateTimeValue}
        scheduled={scheduledValue}
      />
    </div>
  )
})
