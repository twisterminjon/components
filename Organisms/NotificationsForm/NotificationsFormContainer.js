import React, { useMemo } from 'react'
import NotificationsForm from './NotificationsForm'
import UsersQl from '../../../services/UsersQl'
import dayjs from 'dayjs'
import PropTypes from 'prop-types'

import ErrorPage from '../../Views/ErrorPage/ErrorPage'

import { useQuery, useMutation } from '@apollo/react-hooks'

const GET_NOTIFICATIONS_STATUS_QUERY = UsersQl.getMe()
const SET_NOTIFICATIONS_STATUS_QUERY = UsersQl.setNotificationsStatus()

NotificationsFormContainer.propTypes = {
  /** Called when the close button is clicked */
  onClose: PropTypes.func.isRequired,

  /** Called after notifications have been paused */
  onPause: PropTypes.func.isRequired,

  /** Called after notifications have been resumed */
  onResume: PropTypes.func.isRequired,
}

export default function NotificationsFormContainer({ onClose, onPause, onResume }) {
  const { data, error } = useQuery(GET_NOTIFICATIONS_STATUS_QUERY, {
    fetchPolicy: 'cache-and-network',
  })

  const [setNotificationsStatus] = useMutation(SET_NOTIFICATIONS_STATUS_QUERY, {
    onCompleted() {
      onClose()
    },
  })

  const { startDateTime, endDateTime } = useMemo(() => {
    let { startDate: startDateTime, endDate: endDateTime, indefinite } = data.me.notificationsStatus

    startDateTime = startDateTime
      ? dayjs(startDateTime)
          .utc(true)
          .format()
      : null
    endDateTime = endDateTime
      ? dayjs(endDateTime)
          .utc(true)
          .format()
      : null

    return {
      startDateTime,
      endDateTime,
      indefinite,
    }
  }, [data.me.notificationsStatus])

  const handlePause = ({ startDateTime, endDateTime }) => {
    const utcStartDateTime = startDateTime ? dayjs(startDateTime).format() : null
    const utcEndDateTime = endDateTime ? dayjs(endDateTime).format() : null

    setNotificationsStatus({
      variables: {
        startDate: utcStartDateTime,
        endDate: utcEndDateTime,
        indefinite: (startDateTime && !endDateTime) || (!startDateTime && endDateTime),
      },
    })

    if (typeof onPause === 'function') {
      onPause({ startDateTime, endDateTime })
    }
  }

  const handleResume = () => {
    setNotificationsStatus({
      variables: {
        startDate: null,
        endDate: null,
        indefinite: false,
      },
    })

    if (typeof onResume === 'function') {
      onResume()
    }
  }

  const paused = data.me.notificationsStatus.paused

  const scheduled = useMemo(() => startDateTime && !paused, [startDateTime, paused])

  if (error) {
    return <ErrorPage error={error} />
  }

  return (
    <NotificationsForm
      onClose={onClose}
      paused={paused}
      defaultPauseStartDateTime={startDateTime}
      defaultPauseEndDateTime={endDateTime}
      onPause={handlePause}
      onResume={handleResume}
      scheduled={scheduled}
    />
  )
}
