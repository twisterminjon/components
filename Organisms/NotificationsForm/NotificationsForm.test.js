import React from 'react'
import NotificationsForm from './NotificationsForm'

describe('NotificationsForm', () => {
  const mockFun = jest.fn()

  it('matches the snapshot', () => {
    const wrapper = window.shallow(
      <NotificationsForm show={true} onClose={mockFun} onPause={mockFun} paused={false} onResume={mockFun} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('matches the snapshot when hidden', () => {
    const wrapper = window.shallow(
      <NotificationsForm show={false} onClose={mockFun} onPause={mockFun} paused={false} onResume={mockFun} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
