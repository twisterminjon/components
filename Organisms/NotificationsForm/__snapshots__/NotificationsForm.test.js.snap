// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsForm matches the snapshot 1`] = `
<div
  className="notificationsform"
  data-testid="notificationsform"
>
  <div
    className="notificationsform-title notificationsform-section"
  >
    Notifications
  </div>
  <hr />
  <div
    className="notificationsform-section"
  >
    <p>
      Pause your notifications.
    </p>
    <p>
      When your notifications are paused other users will be notified that you are unavailable.
    </p>
  </div>
  <hr />
  <div>
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-Until I turn off"
      key="0"
      label="Until I turn off"
      loading={false}
      name="until-off"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-30 minutes"
      key="1"
      label="30 minutes"
      loading={false}
      name="30-minutes"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-1 hour"
      key="2"
      label="1 hour"
      loading={false}
      name="1-hour"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-2 hours"
      key="3"
      label="2 hours"
      loading={false}
      name="2-hours"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-4 hours"
      key="4"
      label="4 hours"
      loading={false}
      name="4-hours"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-Until tomorrow"
      key="5"
      label="Until tomorrow"
      loading={false}
      name="until-tomorrow"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-Custom"
      key="6"
      label="Custom"
      loading={false}
      name="custom"
      onClick={[Function]}
      style={Object {}}
    />
  </div>
  <div
    className="notificationsform-section notificationsform-section--last"
  >
    <Button
      data-testid="notificationsform-close-button"
      fluid={true}
      onClick={[MockFunction]}
      size="big"
      style={Object {}}
      type="submit"
    >
      Close
    </Button>
  </div>
</div>
`;

exports[`NotificationsForm matches the snapshot when hidden 1`] = `
<div
  className="notificationsform"
  data-testid="notificationsform"
>
  <div
    className="notificationsform-title notificationsform-section"
  >
    Notifications
  </div>
  <hr />
  <div
    className="notificationsform-section"
  >
    <p>
      Pause your notifications.
    </p>
    <p>
      When your notifications are paused other users will be notified that you are unavailable.
    </p>
  </div>
  <hr />
  <div>
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-Until I turn off"
      key="0"
      label="Until I turn off"
      loading={false}
      name="until-off"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-30 minutes"
      key="1"
      label="30 minutes"
      loading={false}
      name="30-minutes"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-1 hour"
      key="2"
      label="1 hour"
      loading={false}
      name="1-hour"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-2 hours"
      key="3"
      label="2 hours"
      loading={false}
      name="2-hours"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-4 hours"
      key="4"
      label="4 hours"
      loading={false}
      name="4-hours"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-Until tomorrow"
      key="5"
      label="Until tomorrow"
      loading={false}
      name="until-tomorrow"
      onClick={[Function]}
      style={Object {}}
    />
    <MenuItem
      className=""
      count={null}
      data-testid="notificationsform-menu-item-Custom"
      key="6"
      label="Custom"
      loading={false}
      name="custom"
      onClick={[Function]}
      style={Object {}}
    />
  </div>
  <div
    className="notificationsform-section notificationsform-section--last"
  >
    <Button
      data-testid="notificationsform-close-button"
      fluid={true}
      onClick={[MockFunction]}
      size="big"
      style={Object {}}
      type="submit"
    >
      Close
    </Button>
  </div>
</div>
`;
